main/src/main/scala/sbt/Cross.scala

```diff
@@ -127,76 +128,76 @@ object Cross {
+              "that are configured."
+          )
+          state.log.debug("Scala versions configuration is:")
+          projCrossVersions.foreach {
+            case (project, versions) => state.log.debug(s"$project: $versions")
           }
+        }
 
-          // Execute using a blanket switch
-          projCrossVersions.toMap.apply(currentRef).flatMap { version =>
-            // Force scala version
-            Seq(s"$SwitchCommand $verbose $version!", aggCommand)
+        // Execute using a blanket switch
+        projCrossVersions.toMap.apply(extracted.currentRef).flatMap { version =>
+          // Force scala version
+          Seq(s"$SwitchCommand $verbose $version!", aggCommand)
+        }
+      case Right((keys, taskArgs)) =>
+        def project(key: ScopedKey[_]): Option[ProjectRef] = key.scope.project.toOption match {
+          case Some(p: ProjectRef) => Some(p)
+          case _                   => None
+        }
+        val fullArgs = if (taskArgs.trim.isEmpty) "" else s" ${taskArgs.trim}"
+        val keysByVersion = keys
+          .flatMap { k =>
+            project(k).toSeq.flatMap(crossVersions(extracted, _).map(v => v -> k))
           }
-
-        case Right(_) =>
-          // We have a key, we're likely to be able to cross build this using the per project behaviour.
-
-          // Group all the projects by scala version
-          projVersions.groupBy(_._2).mapValues(_.map(_._1)).toSeq.flatMap {
-            case (version, Seq(project)) =>
-              // If only one project for a version, issue it directly
-              Seq(s"$SwitchCommand $verbose $version $project/$aggCommand")
-            case (version, projects) if aggCommand.contains(" ") =>
-              // If the command contains a space, then the `all` command won't work because it doesn't support issuing
-              // commands with spaces, so revert to running the command on each project one at a time
-              s"$SwitchCommand $verbose $version" :: projects
-                .map(project => s"$project/$aggCommand")
-            case (version, projects) =>
-              // First switch scala version, then use the all command to run the command on each project concurrently
-              Seq(
-                s"$SwitchCommand $verbose $version",
-                projects.map(_ + "/" + aggCommand).mkString("all ", " ", "")
-              )
+          .groupBy(_._1)
+          .mapValues(_.map(_._2).toSet)
+        val commandsByVersion = keysByVersion.toSeq
+          .flatMap {
+            case (v, keys) =>
+              val projects = keys.flatMap(project)
+              keys.toSeq.flatMap { k =>
+                project(k).filter(projects.contains).flatMap { p =>
+                  if (p == extracted.currentRef || !projects.contains(extracted.currentRef)) {
+                    val parts = project(k).map(_.project) ++ k.scope.config.toOption.map {
+                      case ConfigKey(n) => n.head.toUpper + n.tail
```

Comment:
Is this correct?  I'm not sure what the string `n` can be, but will this do the right thing with a configuration defined as `val AppTest = config("app-test")`?

Either way, this logic should exist in some more central location than buried in here.

Reply:
That doesn't look right, and I missed it during the review.
The logic for slash displaying is around here - https://github.com/sbt/sbt/blob/develop/main-settings/src/main/scala/sbt/Def.scala

Reply:
I could not figure out how to get the config to display correctly with any of the Def apis. I had this problem with Continuous as well. Here is an attempt at a more central solution to this problem: https://github.com/sbt/sbt/pull/5332.